jax/experimental/host_callback.py

```diff
@@ -307,20 +307,24 @@ def pp_val(arg) -> ppu.PrettyPrint:
   * nr_untapped: how many positional arguments (from the tail) should not be
   passed to the tap function.
   * arg_treedef: the treedef of the tapped positional arguments.
-  * transforms: a tuple of the transformations that have been applied.
-  * batch_dims: a tuple of the dims that have been batched, for vmap.
-  * logical_shapes: a tuple of evaluated logical shapes, for mask.
+  * transforms: a tuple of the transformations that have been applied. Each 
+    element of the tuple is itself a tuple with the first element the name
+    of the transform. The remaining elements depend on the transform. For 
+    example, for `batch`, the parameters are the dimensions that have been 
+    batched, and for `mask` the logical shapes. 
 
   * the remaining parameters are passed to the tap function.
 """
-# TODO(necula): handle nested vmap and mask
 id_tap_p = core.Primitive("id_tap")
 id_tap_p.multiple_results = True
 xla.outfeed_primitives.add(id_tap_p)
 
-def _add_transform_name(params: Dict, transform: str) -> Dict:
+def _add_transform_name(params: Dict, transform: str,
+                        *transform_params) -> Dict:
   """Adds the `transform` to the params["transforms"]."""
-  return dict(params, transforms=params.get("transforms", ()) + (transform,))
+  new_transform = (transform, *transform_params)
```

Comment:
My inclination would be to always make `new_transform` always of type `Tuple[str, Dict[str, Any]]`, e.g.,
```python
def _add_transform_name(params: Dict, transform: str,
                        **transform_params) -> Dict:
  """Adds the `transform` to the params["transforms"]."""
  new_transform = (transform, transform_params)
```

The current form `Union[Tuple[str], Tuple[str, Dict[str, Any]] is not really an improvement in consistency over your original `Union[str, Tuple[str, Dict[str, Any]]`. Both would need to use a check (`isintance()` or `len()`) to disambiguate the `Union` type.

Reply:
`new_transforms` is `Tuple[str, ...]` (you can see this from the tests that print `transforms`). One can read transform[0] to get the name of the transform, and then you know whether to read more from the tuple. But I can change it as you propose.

Reply:
Actually, I just remembered why I did not use a dictionary: it is not hashable, and we need to hash all the kwargs of id_tap. This is used to get the consumer ID. There are ways around this, e.g., generating a new consumer ID every time, wrapping the Dict under an object. But is it worth it, what do you think?

Reply:
That would be my vote. It's nice to have a consistent type signature given that the set of JAX transforms will presumably increase over time, e.g., only the most expert users would even know `mask` exists currently.

Reply:
If you want a hashable key, you could use `tuple(params.items())` perhaps?

I was mostly thinking of how the expertise would be for users of `id_tap`...

Reply:
I found a compromise. The internal representation is as Tuple[str, ...]. But when I pass it on to the user-function I unpack it into a dict(name="batch", batch_dims=(0, 0)). So, `transforms` is a tuple of dict, always.